dbt/adapters/duckdb/impl.py

```diff
@@ -176,6 +187,28 @@ def get_rows_different_sql(
         )
         return sql
 
+    @available.parse(lambda *a, **k: [])
+    def get_column_schema_from_query(self, sql: str) -> List[Column]:
+        """Get a list of the Columns with names and data types from the given sql."""
+        _, cursor = self.connections.add_select_query(sql)
```

Comment:
Going to talk out loud here a bit-- I've been reading through this code and the code in dbt-core to understand how this is intended to work and how I'm going to square it with DuckDB and with my future plan for [environments](https://github.com/jwills/dbt-duckdb/blob/master/dbt/adapters/duckdb/environments.py) to enable both local and remote execution of models in dbt-duckdb.

So I'm thinking that I'm going to override both the `add_select_query` and the `data_type_code_to_name` method on the `DuckDBConnectionManager` class, and they will in turn delegate their work to the (singleton) `Environment` instance associated that is associated with the run. *The consequence of making that choice is that, by overriding those two methods, we should _not_ have to override the impl of `get_column_schema_from_query` in the adapter impl as we are here.*

Right now, I have two environment impls in the repo: a local one (the default), and a Buena Vista one that uses the Postgres protocol to talk to a Python server running DuckDB. As we discussed, Buena Vista already does (some of) the work required to translate the data types returned from DuckDB into Postgres-compatible type descriptions using the Arrow interface, which provides more detailed type information than the DBAPI implementation in DuckDB. I will need to flesh out those conversions a bit more so as to support all of the conversions specified in the `data_types` fixture in the adapter functional tests, and then copy an edited version of that code from the Buena Vista repo into dbt-duckdb repo for the `LocalEnvironment` implementation to use. I'm not wild about doing this kind of horrible copy-pasta, but it's the only way I can see having both 1) a proper implementation of model contracts in dbt-duckdb and 2) keeping the environments construct that I have totally fallen in love with in-place. The right longer term solution is going to be better/richer type information included in the `description` field returned by DuckDB's DBAPI implementation, but that's going to be months away in the best case, and patience is not one of my virtues.

Additionally, there is some more source-side plugin work I need to complete (I'm slowly coming to terms with the fact that the beautiful dream of write-side plugins in #143 isn't going to happen until at least 1.5.1) and some other stuff I have going on this week that is going to make completing that body and having everything tested to my satisfaction by Friday is going to be difficult to pull off.

Reply:
There might be another (more naive) approach, if need be. Rather than trying to pull this off with a single (empty) query, and pull the data types off the cursor `description`, we could opt for two queries instead:
- create a temp table from the empty query
- `adapter.get_columns_in_relation` (a.k.a. `information_schema.columns`) - which should return more precise data types

Something like:
```sql
{% macro get_column_schema_from_query(select_sql) -%}
    {% set tmp_relation = ... %}
    {% call statement('create_tmp_relation') %}
       {{ create_table_as(temporary=True, relation=tmp_relation, compiled_code=get_empty_subquery_sql(compiled_code)) }}
    {% endcall %}
    {% set column_schema = adapter.get_columns_in_relation(tmp_relation) %}
    {{ return(column_schema) }}
{% endmacro %}
```
(It looks like we didn't `dispatch` [that macro](https://github.com/dbt-labs/dbt-core/blob/814eb65d59ba71a0a200618021f35429dfba9da6/core/dbt/include/global_project/macros/adapters/columns.sql#L56O) - an oversight that we could quickly correct!)